src/librustdoc/clean/types.rs

```diff
@@ -1966,11 +1967,67 @@ crate struct Static {
 }
 
 #[derive(Clone, PartialEq, Eq, Hash, Debug)]
-crate struct Constant {
-    crate type_: Type,
-    crate expr: String,
-    crate value: Option<String>,
-    crate is_literal: bool,
+crate enum Constant {
+    /// This is the wrapper around `ty::Const` for a non-local constant. Because it doesn't have a
+    /// `BodyId`, we need to handle it on its own.
+    TyConst { type_: Type, expr: String },
+    /// A constant (expression) that’s not an item or associated item. These are usually found
+    /// nested inside types (e.g., array lengths) or expressions (e.g., repeat counts), and also
+    /// used to define explicit discriminant values for enum variants.
+    Anonymous { type_: Type, body: BodyId },
+    /// Inlined constant.
+    Extern { type_: Type, def_id: DefId },
+    /// const FOO: u32 = ...;
+    Local { type_: Type, def_id: DefId, body: BodyId },
+}
+
+impl Constant {
+    crate fn expr(&self, tcx: TyCtxt<'_>) -> String {
+        match self {
+            Self::TyConst { expr, .. } => expr.clone(),
+            Self::Extern { def_id, .. } => print_inlined_const(tcx, *def_id),
+            Self::Local { body, .. } | Self::Anonymous { body, .. } => print_const_expr(tcx, *body),
+        }
+    }
+
+    crate fn value(&self, tcx: TyCtxt<'_>) -> Option<String> {
```

Comment:
What does 'value' mean? Why do only `Extern` and `Local` constants have one?

Reply:
It's the `= "whatever";` part. I took the "value" name from the field of the original `clean::Constant` struct.

Reply:
How is it different from 'expr'?

Reply:
That's actually a good question now that I think about it. :laughing:

Reply:
Can you try switching all uses of `value()` to use `expr()` instead and see if it breaks anything? If not, could you remove this function?

Reply:
It's worth a try.

Reply:
So I think it's worth thinking about it later on because:

```rust
let value = c.value(cx.tcx());
let is_literal = c.is_literal(cx.tcx());
let expr = c.expr(cx.tcx());
if value.is_some() || is_literal {
    write!(w, " = {expr};", expr = Escape(&expr));
} else {
    w.write_str(";");
}

if !is_literal {
    if let Some(value) = &value {
        let value_lowercase = value.to_lowercase();
        let expr_lowercase = expr.to_lowercase();

        if value_lowercase != expr_lowercase
            && value_lowercase.trim_end_matches("i32") != expr_lowercase
        {
            write!(w, " // {value}", value = Escape(value));
        }
    }
}
```

Both `expr` and `value` are used.

Reply:
That doesn't really answer my question though  - why not just unconditionally print the expression, instead of printing it for some types of constants not others?

Reply:
It needs to be investigated. But I don't think it's a good idea to do it here. Should I open an issue about this?

Reply:
Hmm, ok, we can look into it in a follow-up. If you open an issue that would be great.